Add training doc (#220)
Oneflow-Inc/libai: docs/source/tutorials/Training.md

# Training

> To run training, we highly recommend users to use standardized “trainer” in LiBai

Reviewer: trainer好像用了中文的引号, 可以改为 ``trainer``

> To run training, we highly recommend users to use standardized “trainer” in LiBai

Reviewer 2: ```suggestion
To run training, we highly recommend users to use the standardized "trainer" in LiBai.
```

## Trainer Abstraction

> We provide a standardized “trainer” abstraction with a hook system that helps simplify the standard training behavior.

Reviewer: 这里的trainer也是~

> We provide a standardized “trainer” abstraction with a hook system that helps simplify the standard training behavior.

Reviewer 2: ```suggestion
We provide a standardized "trainer" abstraction with a hook system that helps to simplify the standard training behavior.
```

> DefaultTrainer is initialized from lazy config system, used by `tools/train_net.py` and many scripts. It includes many standard default behaviors that one might want to opt in, including default configurations for optimizer, learning rate schedule, logging, evaluation, checkpointing etc.

Reviewer: `want to opt in`这个opt是不是改为一个比较具体的词好一些

Author: 这里借鉴的d2的写法, 暂时没有想到其他更好的表达 https://detectron2.readthedocs.io/en/latest/tutorials/training.html

For simple customizations (e.g. change optimizer, evaluator, LR scheduler, data loader, etc.), modify your `lazycall` method in `config.py` according to your own needs (refer to [Config_System](https://libai.readthedocs.io/en/latest/tutorials/Config_System.html#configs-in-libai)).

## Customize a DefaultTrainer

For complicated customizations, we recommend users to overwrite function in [`DefaultTrainer`](https://github.com/Oneflow-Inc/libai/blob/main/libai/engine/default.py).

> In DefaultTrainer, training consists of `run_step in trainer` and `hooks`, you can modify `trainer` and `hooks` according to your own needs

Reviewer: according to your own needs好像少了一个句号

Reviewer: 是不是可以说一下
`DefaultTrainer can be summarized as the following code`类似的意思

    args = default_argument_parser().parse_args()
    main(args)
```

Using a trainer+hook system means there will always be some non-standard behaviors that cannot be supported, especially in research. For this reason, we intentionally keep the trainer & hook system minimal, rather than powerful.

### Add Hooks in Trainer

For extra tasks during training, check the hook system to see if it’s supported.

[`HookBase`](https://github.com/Oneflow-Inc/libai/blob/ffe5ca0e46544d1cbb4fbe88d9185f96c0dc2c95/libai/engine/trainer.py#L28) in `libai/engine/trainer.py` provides a standard behavior for users to use hook. You can overwirte its function according to your own needs. Please refer to [libai/engine/hooks.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/engine/hooks.py)
```python 
class HookBase:
    def before_train(self):
        """
        Called before the first iteration.
        """

    def after_train(self):
        """
        Called after the last iteration.
        """

    def before_step(self):
        """
        Called before each iteration.
        """

    def after_step(self):
        """
        Called after each iteration.
        """
```

> As an example, to print `iter` in trainer during training:

Reviewer: 这里可以改为

**Example:**
To print ``iter`` in trainer during training:

感觉排版会好看一些

```python
class InfoHook(HookBase):
    def before_train(self):
        logger.info(f"start training at {self.trainer.iter}")

    def after_train(self):
        logger.info(f"end training ad {self.trainer.iter}")

    def after_step(self):
        if self.trainer.iter % 100 == 0:
            logger.info(f"iteration {self.trainer.iter}!")
```

Then you can import your `hook` in `tools/my_train_net.py`

### Modify train_step in Trainer

> In LiBai, `EagerTrainer` and `GraphTrainer` are supported in `libai/engine/trainer.py`. `EagerTrainer` supports `train_step` for `eager` mode while `GraphTrainer` supports `train_step` for `graph` mode, the mode is depended on `graph.enabled` in your `config.py`.

Reviewer 2: ```suggestion
LiBai provides `EagerTrainer` and `GraphTrainer` in `libai/engine/trainer.py` by default. `EagerTrainer` is used in `eager` mode while `GraphTrainer` is used in `graph` mode, and the mode is determined by the `graph.enabled` parameter in your `config.py`.
```

Reviewer 2: 这样如何

> # Training

Reviewer: 感觉可以改为 `Customize Training`

我看mmdet里似乎也有介绍如何Customize各个组件，这个是不是在 graph 下还不太支持

https://mmdetection.readthedocs.io/en/latest/tutorials/customize_runtime.html

Author: mmdet里面是介绍了各个组件怎么registry. 然后在config里面调用.

我感觉我们这个部分的说明已经在lazyconfig里面写好了.  再写一遍其实和lazyconfig的文档有点重复

Author: 这个叫`Training` 感觉主要是参考d2的写法: https://detectron2.readthedocs.io/en/latest/tutorials/training.html

To run training, we highly recommend users to use the standardized `trainer` in LiBai

## Trainer Abstraction

We provide a standardized `trainer` abstraction with a hook system that helps to simplify the standard training behavior.

DefaultTrainer is initialized from lazy config system, used by `tools/train_net.py` and many scripts. It includes many standard default behaviors that one might want to opt in, including default configurations for optimizer, learning rate schedule, logging, evaluation, checkpointing etc.

For simple customizations (e.g. change optimizer, evaluator, LR scheduler, data loader, etc.), modify your `lazycall` method in `config.py` according to your own needs (refer to [Config_System](https://libai.readthedocs.io/en/latest/tutorials/Config_System.html#configs-in-libai)).

## Customize a DefaultTrainer

For complicated customizations, we recommend users to overwrite function in [DefaultTrainer](https://github.com/Oneflow-Inc/libai/blob/main/libai/engine/default.py).

> In DefaultTrainer, training consists of `run_step in trainer` and `hooks`, you can modify `trainer` and `hooks` according to your own needs. 

Reviewer 3: ```suggestion
In `DefaultTrainer`, the training process consists of `run_step` and `hooks` which can be modified according to your own needs. 
```

The following code indicates how `run_step` and `hooks` work during training
```python
class DefaultTrainer(TrainerBase):
    def train(self, start_iter: int, max_iter: int):
        ...

        with EventStorage(self.start_iter) as self.storage:
            try:
                self.before_train() # in hooks
                for self.iter in range(start_iter, max_iter):
                    self.before_step() # in hooks
                    self.run_step() # in self._trainer
                    self.after_step() # in hooks
                self.iter += 1
            except Exception:
                logger.exception("Exception during training:")
                raise
            finally:
                self.after_train() # in hooks

```

> Refer to `tools/train_net.py`, you can rewrite `tools/my_train_net.py` with your modified `_trainer` and `hooks`. We will introduce how to modify `_trainer` and `hooks` in next subsection.

Reviewer 3: ```suggestion
Refer to `tools/train_net.py`, you can rewrite `tools/my_train_net.py` with your modified `_trainer` and `hooks`. We will introduce how to modify them in the next subsection.
```

    args = default_argument_parser().parse_args()
    main(args)
```

Using a trainer+hook system means there will always be some non-standard behaviors that cannot be supported, especially in research. For this reason, we intentionally keep the trainer & hook system minimal, rather than powerful.

### Add Hooks in Trainer

For extra tasks during training, check the hook system to see if it’s supported.

[HookBase](https://github.com/Oneflow-Inc/libai/blob/ffe5ca0e46544d1cbb4fbe88d9185f96c0dc2c95/libai/engine/trainer.py#L28) in `libai/engine/trainer.py` provides a standard behavior for users to use hook. You can overwirte its function according to your own needs. Please refer to [libai/engine/hooks.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/engine/hooks.py)
```python 
class HookBase:
    def before_train(self):
        """
        Called before the first iteration.
        """

    def after_train(self):
        """
        Called after the last iteration.
        """

    def before_step(self):
        """
        Called before each iteration.
        """

    def after_step(self):
        """
        Called after each iteration.
        """
```

> As an example, to print `iter` in trainer during training:

Reviewer: ```suggestion
Depending on the functionality of the hook, users can specify what the hook will do at each stage of the training in ``before_train``, ``after_train``, ``before_step``, ``after_step``. For example, to print `iter` in trainer during training:
```

Reviewer: 感觉这样好些~

```python
class InfoHook(HookBase):
    def before_train(self):
        logger.info(f"start training at {self.trainer.iter}")

    def after_train(self):
        logger.info(f"end training ad {self.trainer.iter}")

    def after_step(self):
        if self.trainer.iter % 100 == 0:
            logger.info(f"iteration {self.trainer.iter}!")
```

Then you can import your `hook` in `tools/my_train_net.py`

> ### Modify train_step in Trainer

Reviewer 3: 感觉需要加一个 get_batch 的介绍，如果他要对数据做额外处理，比如类似 mixup 这种操作，我们推荐他在哪里去做，原则上他可以在 run_step 里面，但是现在有 get_batch 这个接口，他应该可以直接在 get_batch 里面进行处理

LiBai provides `EagerTrainer` and `GraphTrainer` in `libai/engine/trainer.py` by default. `EagerTrainer` is used in `eager` mode while `GraphTrainer` is used in `graph` mode, and the mode is determined by the `graph.enabled` parameter in your `config.py`.

> > more details about `eager` and `graph` mode, please refer to [oneflow doc](https://docs.oneflow.org/en/master/basics/08_nn_graph.html)

Reviewer 3: 差一个句号

> As an example, to add a temp variable for model output in run_step:

Reviewer: ```suggestion
As an example, using a temp variable to keep the model's output in run_step:
```
这样感觉好些~

```python
class MyEagerTrainer(EagerTrainer):

    def __init__(self, model, data_loader, optimizer, grad_acc_steps=1):
        super().__init__(model, data_loader, optimizer, grad_acc_steps)
        self.previous_output = None

    def run_step(self, get_batch: Callable):
        ...
        loss_dict = self.model(**data)
        self.previous_output = loss_dict
        ...
```

> Then you can set your `MyEagerTrainer` as `self.trainer` in `tools/my_train_net.py`

Reviewer: 这个部分是不是也给一个代码例子比较好 🤣

Author: 这个部分在上面给了呀
![image](https://user-images.githubusercontent.com/13843079/160049318-6efba1d4-339e-4acc-b5cf-90fc67f704fa.png)

Reviewer: 奥，OK~

Author: 其实是先给了一个总纲, 
然后再在各个小章节里面写了怎么分别修改`hook`和`trainer`

Reviewer: 可以的~我没问题了~

## Logging of Metrics

During training, trainer put metrics to a centralized [EventStorage](https://libai.readthedocs.io/en/latest/modules/libai.utils.html#module-libai.utils.events). You can use the following code to access it and log metrics to it:

```python
from libai.utils.events import get_event_storage

# inside the model:
if self.training:
  value = # compute the value from inputs
  storage = get_event_storage()

>   storage.put_scalar("some_accuracy", value)

Reviewer 3: 4个空格
